src/sentry/sentry_metrics/consumers/indexer/batch.py

```diff
@@ -98,94 +99,136 @@ def _extract_namespace(self, headers: Headers) -> Optional[str]:
-            self.__message_size_max[use_case_id] = max(
-                len(msg.payload.value), self.__message_size_max[use_case_id]
-            )
-            self.__message_size_sum[use_case_id] += len(msg.payload.value)
-
-            # Ensure that the parsed_payload can be cast back to to
-            # IngestMetric. If there are any schema changes, this check would
-            # fail and ParsedMessage needs to be adjusted to be a superset of
-            # IngestMetric again.
-            _: IngestMetric = parsed_payload
-
-            self.parsed_payloads_by_offset[partition_offset] = parsed_payload
+                parsed_payload = self._extract_message(msg)
+                self._validate_message(parsed_payload)
+                self.parsed_payloads_by_meta[broker_meta] = parsed_payload
+            except Exception:
+                self.invalid_msg_meta.add(broker_meta)
 
-        for namespace, cnt in disabled_msgs_cnt.items():
+        for namespace, cnt in skipped_msgs_cnt.items():
             metrics.incr(
                 "process_messages.namespace_disabled",
                 amount=cnt,
                 tags={"namespace": namespace},
             )
 
+    def _extract_message(
+        self,
+        msg: Message[KafkaPayload],
+    ) -> ParsedMessage:
+        assert isinstance(msg.value, BrokerValue)
+        try:
+            parsed_payload: ParsedMessage = json.loads(
+                msg.payload.value.decode("utf-8"), use_rapid_json=True
+            )
+        except rapidjson.JSONDecodeError:
+            logger.error(
+                "process_messages.invalid_json",
+                extra={"payload_value": str(msg.payload.value)},
+                exc_info=True,
+            )
+            raise
+        try:
+            if self.__input_codec:
+                self.__input_codec.validate(parsed_payload)
+        except ValidationError:
+            if settings.SENTRY_METRICS_INDEXER_RAISE_VALIDATION_ERRORS:
+                raise
+            logger.warning(
+                "process_messages.invalid_schema",
+                extra={"payload_value": str(msg.payload.value)},
+                exc_info=True,
+            )
+        try:
+            parsed_payload["use_case_id"] = use_case_id = extract_use_case_id(
+                parsed_payload["name"]
+            )
+        except ValidationError:
+            logger.error(
+                "process_messages.invalid_metric_resource_identifier",
+                extra={"payload_value": str(msg.payload.value)},
+                exc_info=True,
+            )
+            raise
+
+        self.__message_count[use_case_id] += 1
+        self.__message_size_max[use_case_id] = max(
+            len(msg.payload.value), self.__message_size_max[use_case_id]
+        )
+        self.__message_size_sum[use_case_id] += len(msg.payload.value)
+
+        return parsed_payload
+
+    def _validate_message(self, parsed_payload: ParsedMessage) -> None:
```

Comment:
What is the difference in the `ValueError` being raised in this method compared to the `ValidationError` being raised in _extract_messages? At a high level they seem to point to some data quality issue. Would it then make sense to raise `ValidationError` for them as well?

Reply:
I am a little hesitant to throw ValidationError because they belong in the `sentry_kafka_schemas.codecs` library. I feel like ValueError it not a terrible choice since it's description is
"Raised when an operation or function receives an argument that has the right type but an inappropriate value"

Reply:
Any message which does not conform to certain data quality gates should ideally be dealt with in the same manner to keep consistent behavior of the system. If we raise a `ValueError` here, and lets say we have configured DLQ for `ValidationError`, I would believe that those messages would be dropped rather than sent to the DLQ. 

If we agree on keeping consistent behavior, then I would lean towards the side of raising `ValidationError`

Reply:
Discussed offline that both `ValidationError` and `ValueError` would have the same behavior since we raise `InvalidMessage` for both of those.